lib/SILOptimizer/Mandatory/MandatoryInlining.cpp

```diff
@@ -764,137 +764,89 @@ static SILInstruction *tryDevirtualizeApplyHelper(FullApplySite InnerAI,
   return newApplyAI;
 }
 
-/// Inlines all mandatory inlined functions into the body of a function,
-/// first recursively inlining all mandatory apply instructions in those
-/// functions into their bodies if necessary.
-///
-/// \param F the function to be processed
-/// \param AI nullptr if this is being called from the top level; the relevant
-///   ApplyInst requiring the recursive call when non-null
-/// \param FullyInlinedSet the set of all functions already known to be fully
-///   processed, to avoid processing them over again
-/// \param SetFactory an instance of ImmutableFunctionSet::Factory
-/// \param CurrentInliningSet the set of functions currently being inlined in
-///   the current call stack of recursive calls
-///
-/// \returns true if successful, false if failed due to circular inlining.
-static bool
-runOnFunctionRecursively(SILOptFunctionBuilder &FuncBuilder,
-			 SILFunction *F, FullApplySite AI,
-                         DenseFunctionSet &FullyInlinedSet,
-                         ImmutableFunctionSet::Factory &SetFactory,
-                         ImmutableFunctionSet CurrentInliningSet,
-                         ClassHierarchyAnalysis *CHA) {
-  // Avoid reprocessing functions needlessly.
-  if (FullyInlinedSet.count(F))
-    return true;
-
-  // Prevent attempt to circularly inline.
-  if (CurrentInliningSet.contains(F)) {
-    // This cannot happen on a top-level call, so AI should be non-null.
-    assert(AI && "Cannot have circular inline without apply");
-    SILLocation L = AI.getLoc();
-    assert(L && "Must have location for transparent inline apply");
-    diagnose(F->getModule().getASTContext(), L.getStartSourceLoc(),
-             diag::circular_transparent);
-    return false;
+static size_t collectApplies(SILFunction *F, ClassHierarchyAnalysis *CHA,
+                             DominanceAnalysis *DA,
+                             SmallVector<FullApplySite, 8> &collection) {
+  DominanceInfo *DT = DA->get(F);
+  DominanceOrder domOrder(&F->front(), DT, F->size());
+
+  // First collect all the applies
+  SmallVector<FullApplySite, 12> applyInsts;
+  while (SILBasicBlock *block = domOrder.getNext()) {
```

Comment:
Why do you need to use the dominator info here?

Reply:
We don't. I'll remove it.

Reply:
Hmm. To be honest, I'm not quite sure why reading the blocks in dominance order fixes this, but, when inlining an auto closure with `-enable-ownership-stripping-after-serialization`, the verifier fails because `convert_escape_to_noescape [not_guaranteed] not allowed after mandatory passes`. Using dominance order fixes that.